### PR TITLE
C# Documentation and tests

### PR DESCRIPTION
**Pull request by mylinhdao**
# C# Documentation and tests
Closes: #295 

## Description
- add C# documentation
- add tests for C#
- `??=` counts for complexity metric
- default switch expression arm `_ => expression` will not be counted for complexity
## Definition of Done

A PR is only ready for merge once all the following acceptance criteria are fulfilled:

-   [x] There are automated tests for newly written code and bug fixes
-   [x] Documentation ([README.md](https://github.com/MaibornWolff/metric-gardener/blob/main/README.md), [UPDATE_GRAMMARS.md](https://github.com/MaibornWolff/metric-gardener/blob/main/docs/UPDATE_GRAMMARS.md)) has been updated

### CODE REVIEW

File: docs/C_Sharp.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+-   Null-coalescing operators `??` and `??=`
+-   Loops, including
+    -   `for` loops and `foreach` loops
+    -   `while` loops
+    -   `do-while` loops
+-   `case` labels in switch-statements
+-   Not-default switch expression arms in switch expressions `x switch { value1 => expression1, ...}`
+-   `catch` blocks
+-   logical binary operations `&&` and `||`
+-   pattern combinators `and` and `or` in `is`-expression, switch statements and switch expressions
```
Comment (clemens-mw) [inline docs/C_Sharp.md:20]: why do we count switch statements and not only case labels /expression arms?

File: docs/C_Sharp.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+-   everything counted for the ["functions" metric](#functions)
+
+It does **not** count:
+
+-   case guards with the keyword `when`
```
Comment (clemens-mw) [inline docs/C_Sharp.md:25]: why not?

File: test/metric-end-results/CSharpMetrics.test.ts [hunk 3]
```diff
@@ -39,15 +39,23 @@ describe("C# metric tests", () => {
...
         });
 
         it("should not count any class declaration", () => {
-            testFileMetric("classes.cs", FileMetric.complexity, 1);
+            testFileMetric("classes.cs", FileMetric.complexity, 5);
         });
 
         it("should count switch case labels, but no default labels", () => {
-            testFileMetric("case-statements.cs", FileMetric.complexity, 4);
+            testFileMetric("case-statements.cs", FileMetric.complexity, 5);
```
Comment (clemens-mw) [inline test/metric-end-results/CSharpMetrics.test.ts:58]: why not 3?

**Review by clemens-mw — COMMENTED** [review]
LGTM but #265 should be merged first

File: docs/C_Sharp.md (hunk 2 above)
Comment (mylinhdao) [reply docs/C_Sharp.md:25]: lets discuss if we count case guards for complexity in every language in other issue #88

**Review by clemens-mw — APPROVED** [review]

**Pull request merged**